neptune/query/query.go

```diff
@@ -59,14 +69,42 @@ const (
 		`.property('code_list','%s').as('new')` +
 		`.addE('clone_of').to('old')` +
 		`.select('new')`
-	CountHierarchyNodes         = `g.V().hasLabel('_hierarchy_node_%s_%s').count()`
+
+	// CloneHierarchyNodesFromIDs traverses the provided node IDs and creates a clone for each one
+	// by cloning 'code' and 'label' properties, setting 'hasData' to the provided boolean value, setting 'code_list' to the provided value,
+	// and creating a 'clone_of' edge between the new node and the original one.
+	CloneHierarchyNodesFromIDs = `g.V(%s).as('old')` +
```

Comment:
I think you are right - and if this is always true, we could actually skip the step `GetGenericHierarchyNodeIDs` for "leaf" nodes (or nodes with data), and instead create a list of IDs with 'label-code'!

I didn't assume any relation between codes and node IDs, but if this is always true, that would be a free performance improvement (we may just create the list in hierarchy builder from the codes, instead of calling GetGenericHierarchyNodeIDs)

Reply:
in order to concatenate, we could just just fmt:
`nodeID := fmt.Sprintf("%s--%s", codeListID, code)` . example : `nodeID := fmt.Sprintf("%s--%s", "cpih1dim1aggid", "cpih1dim1S90401")`

Reply:
Note that we would still need to call GetGenericHierarchyNodeIDs for the ancestries

Reply:
The tricky part is that you need to concat the ID together within the query itself, as you are creating each instance hierarchy node within the clone. You would also need to include the instance ID in the node ID, so the ID would be something like `_hierarchy_node_{instanceID}_{codeListID}_{code}`

Reply:
This is called by CloneNodesFromIDs, which receives a map of IDs. The IDs are provided by `GetGenericHierarchyNodeIDs` / `GetGenericHierarchyAncestriesIDs` (which executes `g.V().hasLabel('_generic_hierarchy_node_cpih1dim1aggid') ...` but only to obtain IDs for the provided codes). In the gremlin console (in the develop testing neptune cluster) I see the following:
```gremlin> g.V().hasLabel('_generic_hierarchy_node_cpih1dim1aggid')
==>v[cpih1dim1aggid--cpih1dim1T20000]
==>v[cpih1dim1aggid--cpih1dim1G90100]
==>v[cpih1dim1aggid--cpih1dim1S90501]
==>v[cpih1dim1aggid--cpih1dim1S90201]
...```

Do the generic hierarchy nodes always have this ID format (`cpih1dim1aggid--cpih1dim1S90201`)?

Then I see that the cloned hierarchy nodes have IDs with the format: `_hierarchy_node_{instanceID}_{dimensionName}` What node type has the `_hierarchy_node_{instanceID}_{codeListID}_{code}` format? Maybe I'm missing something?

Reply:
Yes, the same format us always used for the generic hierarchy nodes.

No node type currently has the `_hierarchy_node_{instanceID}_{codeListID}_{code}` format. This is something that I made up when trying to add a unique ID to the cloned hierarchy nodes.

The cloned hierarchy nodes currently use `_hierarchy_node_{instanceID}_{dimensionName}`, though this is a label - not an ID. If we were to create a unique ID property for cloned hierarchy nodes, it would need to be instance specific, hence the `_hierarchy_node_{instanceID}_{codeListID}_{code}`

Reply:
ok! Tahnks! I thought they were IDs